intrusive_collections_examples/src/basic_datastructures/red_black_tree.rs

```diff
@@ -0,0 +1,359 @@
+    /// single root node, it is a complete tree by itself.
+    color: RedBlackTreeNodeColor,
+
+    /// The node that is considered to be the parent of this node.  If there is
+    /// no parent, then this will be `None`.
+    parent: Option<*const RedBlackTreeNode<T>>,
+
+    /// The node that is considered to be the left child of this node.  If there
+    /// is no left child, then this will be `None`.
+    left_child: Option<*const RedBlackTreeNode<T>>,
+
+    /// The node that is considered to be the right child of this node.  If
+    /// there is no right child, then this will be `None`.
+    right_child: Option<*const RedBlackTreeNode<T>>,
+
+    /// Since we're using raw pointers, we **cannot** allow these instances to
+    /// be moved by the compiler (dropping is OK, but moving will almost
+    /// certainly cause UB).  For more information on why this is necessary,
+    /// start reading at
+    /// https://doc.rust-lang.org/std/pin/index.html#example-self-referential-struct
+    /// and be extra careful to read all the way through the explanations about
+    /// drop!
+    _pin: PhantomPinned,
+}
+
+impl<T> Drop for RedBlackTreeNode<T>
+where
+    T: Debug + Clone + Default + PartialEq + Eq + PartialOrd + Ord + Hash,
+{
+    /// Our custom Drop implementation to fix up our neighbors' pointers.
+    ///
+    /// When a node is dropped, it leaves a hole in the tree.  All of the node's
+    /// neighbors (parents and children) will have pointers to a node that no
+    /// longer exists.  We want to fix them up by making them forget about this
+    /// node, if they haven't already forgotten about it.
+    ///
+    /// ## Safety
+    ///
+    /// This is potentially unsafe; there is no guarantee that when we traverse
+    /// the pointers that they are actually pointing to valid memory.  Even if
+    /// they *are* pointing to valid memory, there's no guarantee that we have a
+    /// tree!  If the pointers actually form some kind of arbitrary
+    /// non-tree-like graph, arbitrarily bad things can happen, which is why we
+    /// need to use `unsafe` block below.  In addition, I'm casting away the
+    /// mutability/immutability constraints, which can trick the compiler into
+    /// doing very weird things.
+    fn drop(&mut self) {
+        unsafe {
+            unsafe fn fixup_neighbor<T>(
+                s: &RedBlackTreeNode<T>,
+                neighbor: *mut RedBlackTreeNode<T>,
+            ) where
+                T: Debug
+                    + Clone
+                    + Default
+                    + PartialEq
+                    + Eq
+                    + PartialOrd
+                    + Ord
+                    + Hash,
+            {
+                let self_pointer: *const RedBlackTreeNode<T> = s;
+                if let Some(parent) = (*neighbor).parent {
+                    if parent == self_pointer {
+                        (*neighbor).parent = None;
+                    }
+                }
```

Comment:
Instead of using `if let` followed by an `if`, you can d

```suggestion
                if Some(self_pointer) = (*neighbor).parent {
                    (*neighbor).parent = None;
                }
```

Reply:
Actually, now that I'm looking at this code, I can see that it's quite, quite broken.  The reason for the guards was because `fixup_neighbor()` was called on *all* of `self`'s neighbors, and that `fixup_neighbor()` had to figure out which neighbor was being fixed up.  I'll rewrite it.